Added description of failure scenarios when code uses databases and brokers together (#3609)
MicrosoftDocs/architecture-center: docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md

  - If a message is not successfully processed a configurable number of times, or the handler forwards the message to the dead-letter queue.
- Because messages can potentially be processed more than one time, message handlers should be made idempotent.
  - "A ... method is considered `idempotent` if the intended effect on the server of multiple identical requests with that method is the same as the effect for a single such request." ([RFC 7231](https://tools.ietf.org/html/rfc7231#section-4))
  - One common technique of making message handling idempotent is to check some persistent store, like a database, if the message has already been processed, and if so, not running the logic to process it again. 
  - However, if the processing of the message includes database operations, specifically the insertion of new records with database-generated identifiers, as well as the emitting of new messages to the broker containing those identifiers, because there aren't distributed transactions which encompass both the database and the message broker, there can be a number of complications that can occur if the process running the code happens to fail, for example:

>     - If the code emitting the messages runs before the database transaction is committed, which is how many developers work using the Unit of Work pattern, those messages can "escape" if the failure occurs between calling the broker and asking that the database transaction be committed. As the transaction rolls back, those database-generated IDs are also undone, leaving them available to other code which may be running at the same time. This can cause recipients of the "escaped" messages to processes the wrong database entries, hurting the overall consistency and correctness of your system.

Reviewer: This describes the complications, but does't provide guidance. What's the recommendation here?

Author: Added recommendations.

>     - For a deeper discussion on these failure scenarios and how to handle them, see [this webinar on the topic](https://particular.net/webinars/implementing-an-outbox).

Reviewer: ```suggestion
```

Reviewer: Please remove this. A link to your company's webinar, promoting your product, is not part of Azure Mission-critical guidance.

Author: Done.

  - If an acknowledgment isn't received by the broker in the allotted time period, or the handler explicitly abandons the message, the exclusive lock is released. The message is then available for other consumers to process the message.
  - If a message is not successfully processed a configurable number of times, or the handler forwards the message to the dead-letter queue.
- Because messages can potentially be processed more than one time, message handlers should be made idempotent.

>   - "A ... method is considered `idempotent` if the intended effect on the server of multiple identical requests with that method is the same as the effect for a single such request." ([RFC 7231](https://tools.ietf.org/html/rfc7231#section-4))

Reviewer 2: ```suggestion
  - "A ... method is considered _idempotent_ if the intended effect on the server of multiple identical requests with that method is the same as the effect for a single such request." ([RFC 7231](https://tools.ietf.org/html/rfc7231#section-4))
```

  - However, if the processing of the message includes database operations, specifically the insertion of new records with database-generated identifiers, as well as the emitting of new messages to the broker containing those identifiers, because there aren't distributed transactions which encompass both the database and the message broker, there can be a number of complications that can occur if the process running the code happens to fail, for example:
    - If the code emitting the messages runs before the database transaction is committed, which is how many developers work using the Unit of Work pattern, those messages can "escape" if the failure occurs between calling the broker and asking that the database transaction be committed. As the transaction rolls back, those database-generated IDs are also undone, leaving them available to other code which may be running at the same time. This can cause recipients of the "escaped" messages to processes the wrong database entries, hurting the overall consistency and correctness of your system.

>     - If developers put the code which emits the message *after* the database transaction completes, if the process fails right between the two of them, when the message gets processed again, the idempotence guard clause which checks whether the message was processed before will see that it was (based on the data stored in the database) and then skip the message emitting code, believing that everything was done successfully last time. In this manner, downstream systems which were expecting to receive notifications about the completed process do not receive anything, once again resulting in an overall state of inconsistency.

Reviewer: This is important, but very complicated sentence. I tried to reword it a little but, but also preserve the information:

```suggestion
    - If developers put the code which emits the message *after* the database transaction completes, the process can still fail between these operations (transaction commited - message sent). When that happens, the message will go through processing again, but this time the idempotence guard clause will see that it has been already processed (based on the data stored in the database) and will skip the message emitting code, believing that everything was done successfully last time. Downstream systems which were expecting to receive notifications about the completed process do not receive anything, once again resulting in an overall state of inconsistency.
```

  - The solution to the above problems involves using the [Outbox pattern](/azure/architecture/best-practices/transactional-outbox-cosmos) where the outgoing messages are stored "off to the side" in the same transactional store as the business data and transmitted from there to the message broker when the initial message has been successfully processed.
  - Since many developers are unfamiliar with these problems or their solutions, in order to guarantee that these techniques are applied consistenty in a mission-critical system, it is advisable to have the functionality of the outbox and the interaction with the message broker wrapped in some kind of library, and that all code processing and sending transactionally-significant messages make use of that library rather than interacting with the message broker directly.

>     - Libraries that implement this functionality in .net include [NServiceBus](https://docs.particular.net/nservicebus/outbox/) and [MassTransit](https://masstransit-project.com/advanced/transactional-outbox.html).

Reviewer: ```suggestion
    - Libraries that implement this functionality in .NET include [NServiceBus](https://docs.particular.net/nservicebus/outbox/) and [MassTransit](https://masstransit-project.com/advanced/transactional-outbox.html).
```
